android/greengrass-android-component-base/component-library/src/main/java/com/aws/greengrass/android/component/core/GreengrassComponentService.java

```diff
@@ -0,0 +1,876 @@
+
+        stopForeground(true);
+        stopSelf();
+        setStopped();
+    }
+
+    private void terminateThread(Thread thread, boolean join) {
+        Log.d(LOG_TAG, "terminateThread");
+        if (thread != null && thread.isAlive()) {
+            thread.interrupt();
+            if (join) {
+                try {
+                    thread.join(threadJoinTime.get());
+                } catch (InterruptedException e) {
+                    Log.e(LOG_TAG, "Exception " + Log.getStackTraceString(e));
+                }
+            }
+        }
+    }
+
+    private String getMasterPackage() {
+        return TextUtils.isEmpty(masterPackage) ? NUCLEUS_PACKAGE : masterPackage;
+    }
+
+    private void registerBroadcastReceiver() {
+        registerReceiver(receiver, new IntentFilter(ACTION_STOP_COMPONENT));
+    }
+
+    private class MessengerLooperThread extends Thread {
+        private IncomingSignalsHandler inSigHandler = null;
+        private AtomicReference<Messenger> inSigMessenger = new AtomicReference<>();
+        private Looper inSigLooper = null;
+
+        @Override
+        public void run() {
+            Looper.prepare();
+            synchronized (inSigMessenger) {
+                inSigLooper = Looper.myLooper();
+                inSigHandler = new IncomingSignalsHandler(inSigLooper);
+                inSigMessenger.set(new Messenger(inSigHandler));
+                inSigMessenger.notifyAll();
+            }
+
+            Looper.loop();
+        }
+
+        public IBinder getMessengerBinder() {
+            synchronized (inSigMessenger) {
+                if (inSigMessenger.get() == null) {
+                    try {
+                        inSigMessenger.wait(MESSENGER_READINESS_WAIT_LIMIT_MS);
+                    } catch (InterruptedException ignore) {
+                        Log.d(LOG_TAG, "Interrupted while waiting for the messenger to be ready");
+                    }
+                }
+                Messenger messenger = inSigMessenger.get();
+                if (messenger != null) {
+                    return messenger.getBinder();
+                } else {
+                    return null;
+                }
+            }
+        }
+
+        void finish() {
+            if (this.isAlive()) {
+                // Terminate looper
+                inSigLooper.quitSafely();
+                // Wait for the looper top process all the messages in its queue
+                try {
+                    this.join();
+                } catch (InterruptedException ignore) {
+                    // Nothing to do with it
+                }
+            }
+        }
+    }
+
+    /**
+     * Handler of incoming messages from Nucleus.
+     */
+    private class IncomingSignalsHandler extends Handler {
+
+        public IncomingSignalsHandler(Looper looper) {
+            super(looper);
+        }
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public void handleMessage(Message msg) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.